MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c

```diff
@@ -839,103 +1203,225 @@ NvmeControllerInit (
+    Private->CqData[Index].EntrySize       = Private->CqData[0].EntrySize;
+  }
+
+  // Using the first data queue size for the number of pages required for the data queues
+  QueuePairPageCount = EFI_SIZE_TO_PAGES (Private->SqData[1].NumberOfEntries * (UINTN)LShiftU64 (2, Private->SqData[1].EntrySize))
+                       + EFI_SIZE_TO_PAGES (Private->CqData[1].NumberOfEntries * (UINTN)LShiftU64 (2, Private->CqData[1].EntrySize));
 
+  Status = PciIo->AllocateBuffer (
+                    PciIo,
+                    AllocateAnyPages,
+                    EfiBootServicesData,
+                    QueuePairPageCount * Private->NumberOfDataQueuePairs,
+                    (VOID **)&Private->DataQueueBuffer,
+                    0
+                    );
   if (EFI_ERROR (Status)) {
     return Status;
   }
 
+  Bytes  = EFI_PAGES_TO_SIZE (QueuePairPageCount * Private->NumberOfDataQueuePairs);
+  Status = PciIo->Map (
+                    PciIo,
+                    EfiPciIoOperationBusMasterCommonBuffer,
+                    Private->DataQueueBuffer,
+                    &Bytes,
+                    &MappedAddr,
+                    &Private->DataQueueMapping
+                    );
+
+  if (EFI_ERROR (Status) || (Bytes != EFI_PAGES_TO_SIZE (QueuePairPageCount * Private->NumberOfDataQueuePairs))) {
+    return Status;
+  }
+
+  Private->DataQueueBufferPciAddr = (UINT8 *)(UINTN)MappedAddr;
+
+  Status = NvmeControllerInitIoQueues (Private);
+  // MU_CHANGE [END] - Allocate IO Queue Buffer
+
+  return Status;
+}
+
+// MU_CHANGE [BEGIN] - Allocate IO Queue Buffer
+
+/**
+  Reset the Nvm Express controller.
+
+  @param[in] Private                 The pointer to the NVME_CONTROLLER_PRIVATE_DATA data structure.
+
+  @retval EFI_SUCCESS                The NVM Express Controller is initialized successfully.
+  @retval Others                     A device error occurred while initializing the controller.
+
+**/
+EFI_STATUS
+NvmeControllerReset (
+  IN NVME_CONTROLLER_PRIVATE_DATA  *Private
+  )
+{
+  EFI_STATUS           Status;
+  EFI_PCI_IO_PROTOCOL  *PciIo;
+  UINT32               Index;
+  UINT16               VidDid[2];
+  UINT8                Sn[21];
+  UINT8                Mn[41];
+  NVME_AQA             Aqa;
+  UINT32               NdqpBeforeReset;
+
+  DEBUG ((DEBUG_INFO, "%a: Begin Controller Reset\n", __func__));
+
+  PciIo = Private->PciIo;
+
   //
-  // Program admin completion queue address.
+  // Verify the controller is still accessible
   //
-  Status = WriteNvmeAdminCompletionQueueBaseAddress (Private, &Acq);
+  Status = PciIo->Pci.Read (
+                        PciIo,
+                        EfiPciIoWidthUint16,
+                        PCI_VENDOR_ID_OFFSET,
+                        ARRAY_SIZE (VidDid),
+                        VidDid
+                        );
+  if (EFI_ERROR (Status)) {
+    ASSERT_EFI_ERROR (Status);
+    return EFI_DEVICE_ERROR;
+  }
+
+  if ((VidDid[0] == EFI_PAGE_MASK) || (VidDid[1] == EFI_PAGE_MASK)) {

```

Comment:
I don't know what this is trying to do with vendor id...why is it comparing these to EFI_PAGE_MASK?

Reply:
This comes from the commit here: https://github.com/microsoft/mu_basecore/commit/1e93d34be2ca33aa868460e0b3cda45a43cdabae#diff-e8680836e3bdb35921692e1fb5621eb8bec926e59953e6b9ad249cbfa1e63205R763

If the VidDid is 0xFFFFF it's not valid or wasn't read properly
@os-d suggested using EFI_PAGE_MASK instead of the hard-coded value.

Reply:
That is not what I meant :). This was `0xFFFF` not `0xFFF` for one thing, but also does not relate to the page mask. My comment was for the cases where you were checking something was page aligned

![image](https://github.com/user-attachments/assets/dff884b1-afaa-4174-90e3-63f78ca66c85)

Reply:
Having this compared to a macro makes sense, but let's use the right one :). Which in this case would be creating an `INVALID_VID_DID` or whatever (or if something already exists). But because you are replaying the commit on top, I don't think it needs to change.

I did have a question about the replaying of that commit, I know that came from my comment asking if this should be squashed with that or if it was completely subsuming that, but the answer is no. So does it make it cleaner to do the revert and then the replay? If not, we can just put this one on top, I was just trying to understand, it was hard to tell from the original PR if this was completely changing that original commit or not.

Reply:
Added NVME_INVALID_VID_DID in NvmExpress.h